docs/src/plasma: translation-triggered consistency checks (smoe:docs_plasma) (#1991)
LinuxCNC/linuxcnc: docs/src/plasma/qtplasmac.adoc

== Preamble

*Except where noted, this guide assumes the user is using the latest version of QtPlasmaC.
Version history can be seen by visiting this http://htmlpreview.github.io/?https://github.com/LinuxCNC/linuxcnc/blob/master/share/qtvcp/screens/qtplasmac/versions.html[link] which will show the latest available version.

> The installed QtPlasmaC version is displayed in the title bar. See <<plasma:update,Update QtPlasmaC>> for information on updating QtPlasmaC.*

Author: I suggest to have another round of changes for these kind of improvements.

LinuxCNC provides two configuration wizards which can be used to build a machine configuration. The choice of these wizards is dependent on the hardware used to control the machine.

If the user wishes to use a Run In Place installation then prior to running one of the following commands they will need to run the following command from a terminal: +

> `source ~/linuxcnc-dev/scripts/rip-environment` .

Author: I agree that this would be nice for auto-tests.

LinuxCNC provides two configuration wizards which can be used to build a machine configuration.
The choice of these wizards is dependent on the hardware used to control the machine.

If the user wishes to use a Run In Place installation then prior to running one of the following commands they will need to run the following command from a terminal: +
`source ~/linuxcnc-dev/scripts/rip-environment` .

If using a Package installation then no additional action is required.

> If using a parallel port, use the <<cha:stepconf-wizard,StepConf wizard>> by running the `stepconf` command into a terminal window or launching it using the *Application > CNC > StepConf Wizard* desktop menu entry.

Reviewer: ```suggestion
If using a parallel port, use the <<cha:stepconf-wizard,StepConf wizard>> by running the `stepconf` command in a terminal window or launching it using the *Application > CNC > StepConf Wizard* desktop menu entry.
```
"into" sounds weird here for me

> If using a Mesa Electronics board, use the <<cha:pncconf-wizard,PnCconf wizard>> by running the `pncconf` command into a terminal window or launching it using the *Application > CNC > PnCConf Wizard* desktop menu entry.

Reviewer: ```suggestion
If using a Mesa Electronics board, use the <<cha:pncconf-wizard,PnCconf wizard>> by running the `pncconf` command in a terminal window or launching it using the *Application > CNC > PnCConf Wizard* desktop menu entry.
```
same here

Velocity, and Jog Increments whereas in StepConf these are automatically
calculated and set.

.PnCConf QtPlasmaC Options
image::images/qtplasmac_pncconf_screen.png[width=600,align="center"]

.StepConf QtPlasmaC Options
image::images/qtplasmac_stepconf_options.png[width=600,align="center"]

.QtPlasmaC User Buttons
image::images/qtplasmac_pncconf_buttons.png[width=600,align="center"]

.QtPlasmaC THCAD
image::images/qtplasmac_pncconf_thcad.png[width=600,align="center"]

The THCAD screen will only appear if a Plasma Encoder is selected in the card screen.
The the <<plasma:mesa-thcad,dedicated section on Mesa THCAD>> for more information.

When the configuration is complete, the wizard will save a copy of the configuration that may be loaded and edited at a later time, a working QtPlasmaC configuration will be created in the following directory: ~/linuxcnc/configs/_<machine_name>_.

The newly created QtPlasmaC configuration can be run by entering the following command into a terminal window (*change "_<machine_name>_" to the machine name entered into the configuration wizard*):

For a package installation (Buildbot): +

> `linuxcnc&nbsp;~/linuxcnc/configs/_<machine_name>_/_<machine_name>_.ini`

Reviewer: That wouldn't fit the hierarchy of the document

For a run in place installation: +

> `\~/linuxcnc&#8209;dev/scripts/linuxcnc&nbsp;~/linuxcnc/configs/_<machine_name>_/_<machine_name>_.ini`

Author: This will not work because of the --- blocks not interpreting the _.._ syntax. Otherwise I like this.

If any Qt dependency errors are encountered while attempting to run the QtPlasmaC configuration, the user may need to run the QtVCP installation script to resolve these issues.

For a package installation (Buildbot) enter the following command in a terminal window: +

> `/usr/lib/python3/dist-packages/qtvcp/designer/install_script` .

Author: Avoiding blocks for single lines.

For a run in place installation enter the following command in a terminal window: +

> `~/linuxcnc-dev/lib/python/qtvcp/designer/install_script` .

Author: Avoiding blocks for single lines.

Comment=
Terminal=false
Name=LinuxCNC

> Exec=sh -c "linuxcnc $HOME/linuxcnc/configs/_<machine_name>_/_<machine_name>_.ini"

Reviewer: I think so too. Let's keep it simple:
```suggestion
Exec=sh -c "linuxcnc $HOME/linuxcnc/configs/<machine_name>/<machine_name>.ini"
```

A shutdown warning can be displayed on every shutdown by checking the *Exit Warning* checkbox on the <<qt_parameters-tab,PARAMETERS Tab>>.

[[plasma:main-tab]]
=== MAIN Tab

Screenshot example of the QtPlasmaC <<plasma:main-tab,MAIN Tab>> in *16:9* aspect ratio:

image::images/qtplasmac_16x9.png[width=800,align="center"]

Some functions/features are only used for particular modes and are not displayed if they are not required by the chosen QtPlasmaC mode.

> .Features of the *PREVIEW WINDOW*

Reviewer 2: ```suggestion
.*PREVIEW WINDOW* Features
```

Reviewer: I like that suggestion

    Clicking the label will return the slider to the default linear velocity as set in the <machine_name).ini file.
|===

> .*BUTTONS*  The Button Panel contains buttons useful for the operation of the machine.

Reviewer 2: ```suggestion
.*BUTTONS*
```

Stating the obvious, don't you think ?

Reviewer: Yeah a bit, but I would keep it.

If a full table is displayed due to no G-code file being loaded and the user wishes to change the view orientation, then pressing either Z or P will change the display to the newly selected view. If the user then wishes to display the full table while maintaining the currently selected view as the default view for a loaded G-code file, then pressing CLEAR will achieve this and allow the selected view orientation to prevail the next time a G-code file is loaded.

[[plasma:conversational-tab]]
=== CONVERSATIONAL Tab

> Screenshot example of the QtPlasmaC <<plasma:conversational-tab,CONVERSATIONAL Tab>> in *16:9* aspect ratio:
> 
> image::images/qtplasmac_conversational.png[width=800,align="center"]

Reviewer 2: ```suggestion
.QtPlasmaC *CONVERSATIONAL* Tab in 16:9 aspect ratio
image::images/qtplasmac_conversational.png["QtPlasmaC CONVERSATIONAL Tab in 16:9 aspect ratio"width=800,align="center"]
```

|RELOAD |This button will reload the currently loaded G-code File.
|===

> .*THC*

Reviewer: Why did that change?

Author: Some merge gone wild, I presume. Thank you for spotting that. .THC was above.

setp qtplasmac.conv_disable 1
----

[[plasma:parameters-tab]]
=== PARAMETERS Tab

> Screenshot example of the QtPlasmaC <<plasma:parameters-tab,PARAMETERS Tab>> in *16:9* aspect ratio:
> 
> image::images/qtplasmac_parameters.png[width=800,align="center"]

Reviewer 2: ```suggestion
.QtPlasmaC *PARAMETERS* Tab 16:9 aspect ratio
image::images/qtplasmac_parameters.png["QtPlasmaC PARAMETERS Tab 16:9 aspect ratio"width=800,align="center"]
```

[[plasma:backup]]
The **BACKUP CONFIG** button will create a complete machine configuration backup for archiving or to aid in fault diagnosis.
A compressed backup of the machine configuration will be saved in the user's Linux home directory.
The file name will be __<machine_name>___<version>_<date>_<time>.tar.gz where _<machine_name>_ is the machine name entered in the configuration wizard, <version> is the current QtPlasmaC version the user is on, <date> is the current date (YY-MM-DD), and <time> is the current time (HH-MM-SS).

Prior to the backup being made, the machine log will be saved to a file in the configuration directory named machine_log_<date>_<time>.txt where <date> and <time> are as described above. This file along with up to five previous machine logs will also be included in the backup.

These files are not required by QtPlasmaC and are safe to delete at any time.

[[plasma:statistics-tab]]
=== STATISTICS Tab

> The <<plasma:statistics-tab, STATISTICS Tab>> provides statistics to allow for the tracking of consumable wear and job run times.

Reviewer 2: ```suggestion
.QtPlasmaC *STATISTICS* Tab
image::images/qtplasmac_stats.png["QtPlasmaC STATISTICS Tab",width=800,align="center"]

The <<plasma:statistics-tab, STATISTICS Tab>> provides statistics to allow for the tracking of consumable wear and job run times.
```

The *RS485 PMX STATISTICS* panel will be only be displayed if the user has Hypertherm PowerMax communications and a valid RS485 connection to the PowerMax is established.
This panel will show the *ARC ON TIME* for the PowerMax in hh:mm:ss format.

The *MACHINE LOG* is also displayed on the <<plasma:statistics-tab,STATISTICS Tab>>, this log will display any errors and/or important information that occurs during the current LinuxCNC session.
If the user makes a backup of the configuration from the <<plasma:parameters-tab,PARAMETERS Tab>> then the machine log is also included in the backup.

> image::images/qtplasmac_stats.png[width=800,align="center"]

Reviewer 2: ```suggestion
```

It is important to thoroughly understand the difference between *Synchronized with motion* and *Immediate*:

* *M62 and M63* (Synchronized with Motion) - The actual change of the specified output (P2 (THC) for example) will happen at the beginning of the next motion command.
  If there is no subsequent motion command, the output changes will not occur.
  It is best practice to program a motion code (G0 or G1 for example) right after a M62 or M63.
* *M64 and M65* (Immediate) - These commands happen immediately as they are received by the motion controller. Since these are not synchronized with motion, they will break blending.
  This means if these codes are used in the middle of active motion codes, the motion will pause to activate these commands.

This mode may also be used in conjunction with *Mesh Mode* if the user doesn't require an Arc OK signal to begin the cut.

Both *Mesh Mode* and *Ignore Arc OK* can be enabled/disabled at any time during a job.

[[plasma:cut-recovery]]
=== Cut Recovery

> image::images/qtplasmac_cut_recovery.png[width=250,align="center"]

Reviewer 2: ```suggestion
.QtPlacmaC Cut Recovery panel
image::images/qtplasmac_cut_recovery.png["QtPlacmaC Cut Recovery panel",width=250,align="center"]
```

If a laser offset is in effect when *CANCEL MOVE* is pressed then this offset will also be cleared.

[NOTE]
Cut recovery movements will be limited to a radius of 10&thinsp;mm (0.4") from either the point the program was paused, or from the last point on the cut path if paused motion was used.

IMPORTANT: PUDDLE JUMP IS DISABLED DURING CUT RECOVERY

[[plasma:run-from-line]]
=== Run From Line

> 

Reviewer 2: ```suggestion
```

It is important to note that G-code programs can be run from any selected line using this method, however a leadin may not be possible depending on the line selected. In this case, an error message will be displayed to let the user know the leadin calculation was not possible.

Once the user has selected the starting place, the *CYCLE START* button will blink *"SELECTED _nn_"* where _nn_ is the corresponding line number selected. Clicking this button will bring up the following Run From Line dialog box:

It is not possible to use Run From Line from within a subroutine. If the user selects a line within a subroutine and clicks *"SELECTED _nn_"* then an error message will be displayed that includes the O-Code name of the subroutine.

It is not possible to use Run From Line if previous G-code has set cutter compensation active. If the user selects a line while cutter compensation is active and clicks *"SELECTED _nn_"* then an error message will be displayed.

It is possible to select a new line while Run From Line is active.

> image::images/qtplasmac_run_from_line.png[width=350,align="center"]

Reviewer 2: ```suggestion
.QtPlasmaC Run From Line
image::images/qtplasmac_run_from_line.png["QtPlasmaC Run From Line",width=350,align="center"]
```

*In addition, pressing RETURN (or ENTER) with no entry in the MDI will close the MDI window.*

[[plasma:shape-library]]
== Conversational Shape Library

> image::images/qtplasmac_conversational.png[width=800,align="center"]

Reviewer 2: ```suggestion
.QtPlasmaC Conversational Shape Library
image::images/qtplasmac_conversational.png["QtPlasmaC Conversational Shape Library",width=800,align="center"]
```

> image::images/qtplasmac_conv_block.png[width=100,align="left"]

Reviewer 2: ```suggestion
.QtPlasmaC Conversational Block
image::images/qtplasmac_conv_block.png["QtPlasmaC Conversational Block",width=100,align="left"]
```

In addition, QtPlasmaC will alert the user that an error has been sent to the machine log by displaying the message *"ERROR SENT TO MACHINE LOG"* in the lower left portion of the status bar.

The user may opt to disable the Operator Error popup window,
and view the error messages by going to the <<plasma:statistics-tab,STATISTICS Tab>>

> by changing the following preference to *False* in the *[SCREEN_OPTIONS]* of the _<machine_name>_.prefs file in the _<machine_name>_ directory:

Reviewer 2: ```suggestion
by changing the following preference to `False` in the `[SCREEN_OPTIONS]` of the _<machine_name>_.prefs file in the _<machine_name>_ directory:
```

----
desktop_notify
----

[NOTE]
_<machine_name>_.prefs must be edited with QtPlasmaC closed or any changes will be overwritten on exit.

> Additionally, it is possible for *ERROR SENT TO MACHINE LOG* to flash to get the user's attention by adding or editing the following line in the [GUI_OPTIONS] section of the _<machine_name>_.prefs file:

Reviewer 2: ```suggestion
Additionally, it is possible for *ERROR SENT TO MACHINE LOG* to flash to get the user's attention by adding or editing the following line in the `[GUI_OPTIONS]` section of the _<machine_name>_.prefs file:
```

Port = /dev/ttyusb0
----

If the user is unsure of the name of the port, there is a Python script in the configuration directory
that will show all available ports and can also be used to test communications with the plasma unit prior to enabling this feature in the QtPlasmaC GUI.

> To use the test script follow these instructions:
> 

Reviewer 2: ```suggestion
```

> For a package installation (Buildbot) enter the following command in a terminal window: +
> `pmx485-test`

Reviewer 2: ```suggestion
.Running the PowerMax communication test script on a package install
----
pmx485-test
----
```

> For a run in place installation enter the following two commands in a terminal window: +
> `source ~/linuxcnc-dev/scripts/rip-environment +
> pmx485-test`

Reviewer 2: ```suggestion
.Running the PowerMax communication test script on a run in place installation
----
source ~/linuxcnc-dev/scripts/rip-environment
pmx485-test
----
```

> All language files are kept in ~/linuxcnc-dev/share/screens/qtplasmac/languages.

Reviewer 2: ```suggestion
All language files are kept in `~/linuxcnc-dev/share/screens/qtplasmac/languages`.
```

> The qtplasmac.py file is a Python version of the GUI file used for translation purposes.

Reviewer 2: ```suggestion
The `qtplasmac.py` file is a Python version of the GUI file used for translation purposes.
```

> The .ts files are the translation source files for the translations.

Reviewer 2: ```suggestion
The TS files (`.ts` extension) are the translation source files for the translations.
```

> These are the files that require creating/editing for each language.

Reviewer 2: ```suggestion
These are the files that must be created and edited for each language.
```

The default locale for QtPlasmaC is "_en" which means that any translation files created as "qtplasmac_en.*" will not be used for translations.

> If any of the required utilities (pyuic5, pylupdate5, linguist) are not installed then the user will need to install pyqt5-dev-tools: +

Reviewer 2: ```suggestion
.Creating and editing a translation file for QtPlasmaC

*Prerequisite* +
If any of the required utilities (`pyuic5`, `pylupdate5`, `linguist`) are not installed then the user will need to install `pyqt5-dev-tools`:
```

> `sudo apt install pyqt5-dev-tools`

Reviewer 2: ```suggestion

----
sudo apt install pyqt5-dev-tools
----
```

> Change to the languages directory: +
> `cd ~/linuxcnc-dev/share/qtvcp/screens/qtplasmac/languages`

Reviewer 2: ```suggestion
. Change to the languages directory
+
----
cd ~/linuxcnc-dev/share/qtvcp/screens/qtplasmac/languages
----
```

> If any text changes have been made to the GUI then run the following to update the GUI Python file: +
> `pyuic5 ../qtplasmac.ui > qtplasmac.py`

Reviewer 2: ```suggestion
. If any text changes have been made to the GUI, update the GUI Python file
+
----
pyuic5 ../qtplasmac.ui > qtplasmac.py
----
```

> 
> The user can either create a new translation source file for a non-existing language translation
> or modify an existing translation source file due to changes being made to some text in a QtPlasmaC source file.
> If modifying an existing translation that has had no source file changes then this step is not required.

Reviewer 2: ```suggestion
+
The user can either create a new translation source file for a non-existing language translation
or modify an existing translation source file due to changes being made to some text in a QtPlasmaC source file.
If modifying an existing translation that has had no source file changes then this step is not required.
```

> Create or edit a .ts file: +
> `langfile xx`

Reviewer 2: ```suggestion
. Create or edit a translation file
+
----
langfile xx
----
// FIXME Where _xx_ is what ?
```

Author: Hm. Not too much of a fan.

> 
> [NOTE]
> this command is a script which runs the following: $ pylupdate5 *.py ../*.py ../../../../../lib/python/qtvcp/lib/qtplasmac/*.py -ts qtplasmac_xx.ts

Reviewer 2: ```suggestion
+
[NOTE]
this command is a script which runs the following: $ pylupdate5 *.py ../*.py ../../../../../lib/python/qtvcp/lib/qtplasmac/*.py -ts qtplasmac_xx.ts
```

> The editing of the translation is done with the linguist application: +
> `linguist`

Reviewer 2: ```suggestion
. Launch the Qt Linguist application
+
----
linguist
----
```

Author: I do not really like those blocks since everything in there is verbatim. You have no other style options like slanting arguments etc.

> Then open the required translation file.

Reviewer 2: ```suggestion
. Open the TS file and translate the strings
```
